From 0481e0a7ce10440bcdc699903889c16b227e479d Mon Sep 17 00:00:00 2001
From: Ahmed Ismail <Ahmed.Ismail@arm.com>
Date: Mon, 1 Jul 2024 14:49:50 +0100
Subject: [PATCH 2/3] closefile-validsignature-test: Fix test bugs

This test has two main issues:

* otaPal_CloseFile_ValidSignature test has a hardcoded
filename ("test_happy_path_image.bin") which causes
test failure with custom OTA PAL implementation. The
hardcoded filename must be made configurable at
which point it should be renamed to "non_secure image".

* The RSA Sha256 signature is not actually a valid
signature, so a new signature was generated using
different key pair as the library does not provide
the private key used during the signature generation,
the public key used is added to the test_files directory.

Signed-off-by: Ahmed Ismail <Ahmed.Ismail@arm.com>
---
 .../aws_test_ota_pal_rsa_sha256_signature.h   | 56 +++++++++++--------
 src/ota/ota_pal_test.c                        |  7 ++-
 .../test_files/tfm-rsa-sha256-signer.crt.pem  | 11 ++++
 3 files changed, 50 insertions(+), 24 deletions(-)
 create mode 100644 src/ota/test_files/tfm-rsa-sha256-signer.crt.pem

diff --git a/src/ota/aws_test_ota_pal_rsa_sha256_signature.h b/src/ota/aws_test_ota_pal_rsa_sha256_signature.h
index d4f76fe..d67167f 100644
--- a/src/ota/aws_test_ota_pal_rsa_sha256_signature.h
+++ b/src/ota/aws_test_ota_pal_rsa_sha256_signature.h
@@ -63,30 +63,40 @@ static const int ucInvalidSignatureLength = 256;
  */
 static const uint8_t ucValidSignature[] =
 {
-    0xa9, 0x5f, 0xa1, 0x5d, 0x9b, 0x42, 0x09, 0x78, 0xaf, 0x95, 0xe4, 0x42,
-    0xac, 0xf4, 0xd7, 0x83, 0xea, 0xa9, 0x67, 0x54, 0x32, 0x62, 0x54, 0x21,
-    0x46, 0xf5, 0x91, 0xe7, 0x76, 0xdf, 0x2f, 0xa0, 0x6a, 0xe3, 0xb1, 0xd2,
-    0xda, 0x1b, 0x6d, 0x54, 0x6b, 0xac, 0xb6, 0x38, 0x1d, 0x3c, 0xf1, 0xcf,
-    0x0d, 0x33, 0x84, 0xac, 0x5d, 0xf1, 0x20, 0x7b, 0xfe, 0x2f, 0x9d, 0xf9,
-    0x93, 0x5e, 0x4d, 0x34, 0xdc, 0x21, 0x1c, 0x1b, 0x90, 0xf8, 0x37, 0xc6,
-    0x7d, 0x16, 0x16, 0x25, 0xe3, 0xcc, 0x06, 0xb5, 0xd5, 0xbf, 0x64, 0x64,
-    0x44, 0xa0, 0x60, 0xfe, 0xf3, 0xba, 0xf2, 0xdc, 0xc7, 0x57, 0xbe, 0xca,
-    0xcc, 0x6b, 0x57, 0xc2, 0x89, 0x7c, 0xef, 0xdc, 0xc8, 0xf8, 0xb3, 0x34,
-    0x8f, 0x67, 0x7a, 0x04, 0x3e, 0x9f, 0x96, 0x17, 0x7f, 0x98, 0x6c, 0x87,
-    0x21, 0x70, 0x9d, 0xf8, 0xf4, 0x26, 0xe4, 0xc8, 0x61, 0x57, 0x2a, 0xea,
-    0x08, 0xe1, 0x51, 0x52, 0xbb, 0x53, 0x11, 0x35, 0x1e, 0x39, 0x1a, 0x11,
-    0x2f, 0xc4, 0xa6, 0x19, 0x5e, 0x6a, 0x62, 0xa1, 0x77, 0x7f, 0x6b, 0xc4,
-    0x0a, 0x61, 0xf8, 0x9a, 0x95, 0xd2, 0xfa, 0x09, 0x38, 0xd1, 0xb1, 0xfb,
-    0x75, 0xd1, 0x36, 0xfe, 0x2f, 0xc5, 0x08, 0xc1, 0x35, 0x66, 0x7a, 0x97,
-    0x46, 0x61, 0x8a, 0xfa, 0xbb, 0x83, 0x63, 0x8c, 0xe9, 0x0b, 0x2a, 0x3e,
-    0x88, 0xaa, 0x79, 0x63, 0xb6, 0x92, 0x22, 0xb1, 0x75, 0xeb, 0x20, 0x0d,
-    0x06, 0xb2, 0x01, 0x35, 0x01, 0xa2, 0xa7, 0x21, 0x3f, 0xe9, 0x80, 0x83,
-    0x87, 0xfd, 0x29, 0x54, 0x77, 0x3a, 0x13, 0xa6, 0x10, 0xa7, 0x47, 0xed,
-    0x9f, 0xb1, 0x52, 0xb6, 0x86, 0x54, 0xbc, 0x89, 0x40, 0xa8, 0x74, 0x15,
-    0xd8, 0xa7, 0x69, 0x9b, 0x27, 0x7a, 0x99, 0xde, 0x9d, 0xdc, 0x69, 0xb6,
-    0x0c, 0x57, 0x68, 0x3d
+  0x82, 0xec, 0x6b, 0x31, 0xed, 0xed, 0xbb, 0x33, 0x58, 0x1e, 0x37, 0x43,
+  0x2c, 0x2e, 0xef, 0xc8, 0x2a, 0x5d, 0x98, 0x8d, 0xf8, 0xc6, 0x94, 0xf8,
+  0xd2, 0xb7, 0x19, 0x2b, 0xc2, 0x04, 0xe8, 0x67, 0xf8, 0xfb, 0x1d, 0x7e,
+  0xc4, 0x5c, 0xde, 0x6b, 0x05, 0xbb, 0x08, 0x3e, 0xb0, 0x57, 0x4e, 0x44,
+  0x35, 0xe3, 0x09, 0x22, 0x82, 0xda, 0x55, 0x61, 0x05, 0x8c, 0x53, 0xb1,
+  0x02, 0xc8, 0x54, 0x33, 0xc1, 0x2e, 0x1e, 0x78, 0x9a, 0x21, 0xb9, 0xc5,
+  0xba, 0xe0, 0x56, 0x3f, 0xaa, 0x64, 0x65, 0x36, 0x8d, 0x33, 0xbb, 0x59,
+  0xc2, 0x5f, 0x35, 0xfb, 0x97, 0xaa, 0x2f, 0xd0, 0xab, 0xd3, 0x06, 0xdf,
+  0xca, 0x43, 0x39, 0x6c, 0xff, 0xe9, 0x3b, 0x3f, 0x94, 0xf1, 0xb4, 0xbc,
+  0x8c, 0x31, 0xc1, 0xf9, 0xd0, 0xda, 0x27, 0x5c, 0x03, 0xa8, 0x1f, 0x3d,
+  0x81, 0xc9, 0xb2, 0xd7, 0x0c, 0x64, 0x99, 0x31, 0x32, 0x8b, 0x37, 0x94,
+  0xd6, 0x0e, 0x7e, 0x91, 0xa2, 0xd8, 0x0d, 0xc2, 0xb0, 0x17, 0x26, 0xf9,
+  0x33, 0x9c, 0xed, 0x5c, 0x83, 0x6a, 0x2a, 0x16, 0x89, 0x54, 0xba, 0x5e,
+  0x1c, 0xe3, 0x3e, 0xd8, 0x58, 0x03, 0x9e, 0xe3, 0x7a, 0xc2, 0x1c, 0x6b,
+  0xe8, 0xaf, 0x63, 0x15, 0x4e, 0xe2, 0xe6, 0xe3, 0x79, 0x8b, 0x18, 0x95,
+  0x91, 0x1d, 0x63, 0x10, 0x8d, 0x90, 0x82, 0xb3, 0x8d, 0x1a, 0x05, 0xdc,
+  0x77, 0xc9, 0xe0, 0x23, 0x60, 0xc8, 0x53, 0xd8, 0x02, 0xae, 0xb4, 0xc9,
+  0x50, 0xf6, 0x48, 0x2f, 0x20, 0x50, 0xd0, 0x4e, 0x16, 0xb6, 0x6f, 0x52,
+  0xad, 0x0d, 0xe1, 0x09, 0x07, 0xee, 0x68, 0x7b, 0x06, 0x7e, 0xb7, 0xec,
+  0x38, 0xed, 0x27, 0xeb, 0x75, 0x7d, 0xff, 0xee, 0x81, 0xa6, 0xd0, 0xe5,
+  0xbc, 0xe1, 0x68, 0x1c, 0x8f, 0xf4, 0xac, 0x11, 0x0d, 0x61, 0xff, 0xd2,
+  0x37, 0xb7, 0xed, 0x73, 0x01, 0xf3, 0x9e, 0xd3, 0xd5, 0x75, 0xc0, 0xf6,
+  0x40, 0x88, 0x45, 0xc6, 0x5b, 0x61, 0xcf, 0x5a, 0x08, 0x19, 0xb8, 0x0d,
+  0x42, 0x5f, 0x87, 0x11, 0x69, 0xbb, 0x34, 0xb0, 0x67, 0xcf, 0x00, 0xf7,
+  0x16, 0x4c, 0xb4, 0xc8, 0xfd, 0x4a, 0x46, 0xdc, 0x0c, 0xdb, 0x32, 0xd2,
+  0x3c, 0x21, 0xd4, 0x65, 0xb8, 0xf8, 0x92, 0x34, 0xa5, 0xfa, 0xa6, 0x69,
+  0x0c, 0x70, 0x18, 0x32, 0x27, 0x63, 0x2f, 0x2b, 0xa1, 0xab, 0x05, 0xf5,
+  0x6e, 0x2f, 0x4c, 0x1e, 0x6b, 0xe6, 0xd1, 0x57, 0xb5, 0x1a, 0x3e, 0x60,
+  0xbf, 0xee, 0x14, 0xe3, 0x4a, 0x09, 0xaf, 0x2a, 0xee, 0x29, 0x3c, 0xec,
+  0xda, 0xb8, 0xa6, 0x40, 0xad, 0x8d, 0xf9, 0xf3, 0xb1, 0x87, 0xf8, 0x35,
+  0x0c, 0x9c, 0x9a, 0x72, 0x29, 0x67, 0x20, 0x86, 0xf0, 0x0a, 0xfe, 0x36,
+  0x69, 0xe8, 0x92, 0x94, 0xd2, 0x79, 0x15, 0x1f, 0x8c, 0x40, 0x86, 0x6e
 };
-static const int ucValidSignatureLength = 256;
+static const int ucValidSignatureLength = 384;
 
 /**
  * @brief The type of signature method this file defines for the valid signature.
diff --git a/src/ota/ota_pal_test.c b/src/ota/ota_pal_test.c
index e890faa..9b3668b 100644
--- a/src/ota/ota_pal_test.c
+++ b/src/ota/ota_pal_test.c
@@ -137,7 +137,7 @@ TEST( Full_OTA_PAL, otaPal_CloseFile_ValidSignature )
 
     /* We use a dummy file name here because closing the system designated bootable
      * image with content that is not runnable may cause issues. */
-    xOtaFile.pFilePath = ( uint8_t * ) ( "test_happy_path_image.bin" );
+    xOtaFile.pFilePath = ( uint8_t * ) OTA_PAL_FIRMWARE_FILE;
     xOtaFile.fileSize = sizeof( ucDummyData );
     xOtaStatus = otaPal_CreateFileForRx( &xOtaFile );
     TEST_ASSERT_EQUAL( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
@@ -145,6 +145,11 @@ TEST( Full_OTA_PAL, otaPal_CloseFile_ValidSignature )
     /* We still want to close the file if the test fails somewhere here. */
     if( TEST_PROTECT() )
     {
+        /* The blocksRemaining is set to 1 as to update the component state
+         * to PSA_FWU_CANDIDATE where it would query the implementation info
+         */
+        xOtaFile.blocksRemaining = 1;
+
         /* Write data to the file. */
         bytesWritten = otaPal_WriteBlock( &xOtaFile,
                                            0,
diff --git a/src/ota/test_files/tfm-rsa-sha256-signer.crt.pem b/src/ota/test_files/tfm-rsa-sha256-signer.crt.pem
new file mode 100644
index 0000000..10bf6dd
--- /dev/null
+++ b/src/ota/test_files/tfm-rsa-sha256-signer.crt.pem
@@ -0,0 +1,11 @@
+-----BEGIN PUBLIC KEY-----
+MIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEAsuHwWVHsCa/jbi4G5WEv
+C7zqr+aQwB0zWdD5sWyfdV/5RFsjyAOToRBBfMlepzfd2YBufwZgRZrpQCu0osMA
+TQsHzftESR0ODHrkpvQwUDIzoqWvMfKgU5zJbpGHFGkh88WoB2fbKIvCl810tqZ2
+7WiRxPH/L52OYyUPjlSK/JoW0961iJjBTkbHn2FrPYwTG7vAPhaLPqyLJAiYlyQb
+HBZt9yKYegmtiVfB7ULQO0nm3pa4ClB2TgMpBI33e/c+9IrVegb483H6Pt9hL8X4
+dLYZa9R3F4W5ONeFlX0E6ld8CfY91D6JPJRn2CwUgGdJpBJOXJNqfnkKmYR/M7YB
+bFOuGkjNAifz+i8effbfcWdq34Htbujm0wsMR6JzFdLusMVWpnDtt9uaoXuNYH04
+xq7K74Bsv+dLgfpwUl/rqsPK4BTEoayVVPIk+xqdSrYU5fwgeTai7GaiFdy2K8I0
+kLnLZ2Jw0+t2VAWbTkR7cnvqHoVCE1MlBfSOI1ouEaBxAgMBAAE=
+-----END PUBLIC KEY-----
-- 
2.34.1

